src/rmind/components/llm.py

```diff
@@ -1,46 +1,116 @@
+        self.post_norm: nn.LayerNorm = nn.LayerNorm(embedding_dim)  # ffn
+
+        self.mlp: MLPGLU = MLPGLU(
+            dim_model=embedding_dim,
+            dropout=mlp_dropout,
+            activation="gelu",
+            hidden_layer_multiplier=hidden_layer_multiplier,
+        )
+
+    @override
+    def forward(self, x: Tensor, mask: Tensor) -> Tensor:
+        # f
+        residual = x
+        x_norm = self.pre_norm(x)
+        mha, _ = self.mha(x_norm, x_norm, x_norm, attn_mask=mask, need_weights=False)
+        x = residual + self.resid_drop(mha)
+
+        # g
+        residual = x
+        mlp = self.mlp(self.post_norm(x))
+        return residual + mlp
+
+
+class TransformerEncoder(nn.Module):
+    def __init__(  # noqa: PLR0913, PLR0917
+        self,
+        dim_model: int,
+        num_layers: int,
+        num_heads: int,
+        attn_dropout: float = 0.1,
+        resid_dropout: float = 0.1,
+        mlp_dropout: float = 0.1,
+        hidden_layer_multiplier: int = 1,
+    ) -> None:
+        super().__init__()
+        self.layers = nn.ModuleList([  # pyright: ignore[reportUnannotatedClassAttribute]
+            TransformerEncoderBlock(
+                embedding_dim=dim_model,
+                num_heads=num_heads,
+                attn_dropout=attn_dropout,
+                mlp_dropout=mlp_dropout,
+                resid_dropout=resid_dropout,
+                hidden_layer_multiplier=hidden_layer_multiplier,
+            )
+            for _ in range(num_layers)
+        ])
+        # https://github.com/karpathy/nanoGPT/blob/master/model.py#L182
+        self.layer_norm: nn.LayerNorm = nn.LayerNorm(dim_model)
+
+    @override
+    def forward(self, src: Tensor, mask: Tensor) -> Tensor:
+        x = src
+
+        for layer in self.layers:
+            if self.training:
+                x = checkpoint(layer, x, mask, use_reentrant=False)
+            else:
+                x = layer(x, mask)
+
+        return self.layer_norm(x)
+
+
+class MLPGLU(nn.Module):
     def __init__(
         self,
         dim_model: int,
         dropout: float,
-        activation: str,
         hidden_layer_multiplier: int,
         bias: bool = True,  # noqa: FBT001, FBT002
         *_args: Any,
         **_kwargs: Any,
     ) -> None:
         super().__init__()
         dim_mlp = hidden_layer_multiplier * dim_model
-        self.l1 = nn.Linear(in_features=dim_model, out_features=dim_mlp * 2, bias=bias)
-        self.a1 = build_activation(Activation(activation))
-        self.d1 = nn.Dropout(dropout)
-        self.l2 = nn.Linear(in_features=dim_mlp, out_features=dim_model, bias=bias)
-        self.d2 = nn.Dropout(dropout)
+        self.l1: nn.Linear = nn.Linear(
+            in_features=dim_model, out_features=dim_mlp * 2, bias=bias
+        )
+        self.a1: nn.GELU = nn.GELU()
+        self.d1: nn.Dropout = nn.Dropout(dropout)
+        self.l2: nn.Linear = nn.Linear(
+            in_features=dim_mlp, out_features=dim_model, bias=bias
```

Comment:
[nit] imo annotations don't add much value here, can wrap in `@final` if ruff complains